File: Java/UpdateCredentials/src/main/java/com/updatecredentialsample/encryptcredential/helper/AsymmetricHigherKeyEncryptionHelper.java
// ----------------------------------------------------------------------------
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
// ----------------------------------------------------------------------------

package com.updatecredentialsample.encryptcredential.helper;

import java.math.BigInteger;
import java.security.KeyFactory;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.spec.MGF1ParameterSpec;
import java.security.spec.RSAPublicKeySpec;
import java.util.Base64;

import javax.crypto.Cipher;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;

// Refer C# counterpart: 
// https://github.com/microsoft/PowerBI-CSharp/blob/master/sdk/PowerBI.Api/Extensions/AsymmetricHigherKeyEncryptionHelper.cs
public class AsymmetricHigherKeyEncryptionHelper {
	private static byte KeyLength32 = 0;
	private static byte KeyLength64 = 1;

	private static int KEY_LENGTHS_PREFIX = 2;
	private static int MIN_HMAC_KEY_SIZE_BYTES = 32;
	private static int HMAC_KEY_SIZE_BYTES = 64;
	private static int AES_KEY_SIZE_BYTES = 32;

	private static final String TRANSFORMATION = "RSA/ECB/OAEPPadding";
	private static final String ALGORITHM = "RSA";
	private static final String MESSAGE_DIGEST = "SHA-256";
	private static final String MASK_GENERATION_FUNCTION = "MGF1";

	public static String Encrypt(byte[] plainTextBytes, byte[] modulusBytes, byte[] exponentBytes) throws Exception {

		// Generate ephemeral keys for encryption (32 bytes), hmac (64 bytes)
		byte[] keyEnc = GetRandomBytes(AES_KEY_SIZE_BYTES);
		byte[] keyMac = GetRandomBytes(HMAC_KEY_SIZE_BYTES);

		// Encrypt message using ephemeral keys and Authenticated Encryption
		// Symmetric algorithm and encryptor
		byte[] cipherText = AuthenticatedEncryption.Encrypt(keyEnc, keyMac, plainTextBytes);

		// Encrypt ephemeral keys using RSA
		byte[] keys = new byte[KEY_LENGTHS_PREFIX + keyEnc.length + keyMac.length];

		// Prefixing length of Keys. Symmetric Key length followed by HMAC key length
		keys[0] = (byte) KeyLength32;
		keys[1] = (byte) KeyLength64;

		System.arraycopy(keyEnc, 0, keys, 2, keyEnc.length);
		System.arraycopy(keyMac, 0, keys, keyEnc.length + 2, keyMac.length);

		BigInteger modulus = new BigInteger(1, modulusBytes);
		BigInteger exponent = new BigInteger(1, exponentBytes);

		RSAPublicKeySpec keySpec = new RSAPublicKeySpec(modulus, exponent);
		KeyFactory fact = KeyFactory.getInstance(ALGORITHM);
		PublicKey pubKey = fact.generatePublic(keySpec);

		Cipher cipherKey = Cipher.getInstance(TRANSFORMATION);

		OAEPParameterSpec oaepParams = new OAEPParameterSpec(MESSAGE_DIGEST, MASK_GENERATION_FUNCTION, new MGF1ParameterSpec(MESSAGE_DIGEST),
				PSource.PSpecified.DEFAULT);
		cipherKey.init(Cipher.ENCRYPT_MODE, pubKey, oaepParams);

		byte[] encryptedKeys = cipherKey.doFinal(keys);

		// Prepare final payload
		return Base64.getEncoder().encodeToString(encryptedKeys) + Base64.getEncoder().encodeToString(cipherText);
	}

	private static byte[] GetRandomBytes(int size) {

		byte[] data = new byte[size];
		SecureRandom random = new SecureRandom();
		random.nextBytes(data);
		return data;
	}
}

File: Python/Encrypt credentials/Encryption sample/helper/asymmetrichigherkeyencryptionhelper.py
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.

import base64
import os
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from helper.authenticatedencryption import AuthenticatedEncryption


# Refer C# counterpart:
# https://github.com/microsoft/PowerBI-CSharp/blob/master/sdk/PowerBI.Api/Extensions/AsymmetricHigherKeyEncryptionHelper.cs
class AsymmetricHigherKeyEncryptionHelper:

    KEY_LENGTHS_PREFIX = 2
    HMAC_KEY_SIZE_BYTES = 64
    AES_KEY_SIZE_BYTES = 32

    KEY_LENGTH_32 = 0
    KEY_LENGTH_64 = 1

    def encrypt(self, plain_text_bytes, modulus_bytes, exponent_bytes):
        ''' Encrypts the message with RSA, MGF and SHA hashes

        Args:
            plain_text_bytes (bytes): Message to be encrypted
            modulus_bytes (bytes): Modulus bytes returned from GET gateway API
            exponent_bytes (bytes): Exponent bytes returned from GET gateway API

        Returns:
            String: Encrypted credentials
        '''

        # Generate ephemeral random keys for encryption (32 bytes), hmac (64 bytes)
        key_enc = os.urandom(self.AES_KEY_SIZE_BYTES)
        key_mac = os.urandom(self.HMAC_KEY_SIZE_BYTES)

        authenticated_encryption = AuthenticatedEncryption()

        # Encrypt message using ephemeral keys and Authenticated Encryption
        # Symmetric algorithm and encryptor
        cipher_text = authenticated_encryption.encrypt(
            key_enc, key_mac, plain_text_bytes)

        # Encrypt ephemeral keys using RSA
        keys = bytearray(
            [0] * (len(key_enc) + len(key_mac) + self.KEY_LENGTHS_PREFIX))

        # Prefixing length of Keys. Symmetric Key length followed by HMAC key length
        keys[0] = self.KEY_LENGTH_32
        keys[1] = self.KEY_LENGTH_64

        # Copy key enc and key mac into keys array
        keys[2: len(key_enc) + 2] = key_enc[0: len(key_enc)]
        keys[len(key_enc) + 2: len(key_enc) + len(key_mac) + 2] = key_mac[0: len(key_mac)]

        # Convert exponent and modulus byte arrays to integers
        exponent = int.from_bytes(exponent_bytes, 'big')
        modulus = int.from_bytes(modulus_bytes, 'big')

        # Generate public key based on modulus and exponent returned by the API
        public_key = rsa.RSAPublicNumbers(
            exponent, modulus).public_key(default_backend())

        # Encrypt the data
        # Pass padding algorithm, mask generation function and hashing algorithm
        encrypted_bytes = public_key.encrypt(bytes(keys),
                                             padding.OAEP(mgf=padding.MGF1(algorithm=hashes.SHA256()),
                                                          algorithm=hashes.SHA256(),
                                                          label=None))

        # Return final output
        return base64.b64encode(encrypted_bytes).decode() + base64.b64encode(cipher_text).decode()
